### DISCUSSION TITLE
Easy as BluetoothA2DPSink a2dp_sink;

### DISCUSSION THREAD

**Discussion by MichaelPNolan** [pschatzmann/ESP32-A2DP#272] [Q&A] [answered]
I got this one to work on my ESP32-Wroom-32E without much effort, just had to notice that I copied an example with internal dac option configured and remove that.

I'm really interested to see if 2 things can co-exist from the different libraries like streams while this useful Bluetooth sink is also operational. If not it should be a "mode" I can stop and start as needed. So long have I wanted this.
```
#include "BluetoothA2DPSink.h"

BluetoothA2DPSink a2dp_sink;

#define PIN_I2S_BCK   14
#define PIN_I2S_WS    15
#define PIN_I2S_DATA_OUT 27

void setup() {
 static const i2s_config_t i2s_config = {
              .mode = (i2s_mode_t) (I2S_MODE_MASTER | I2S_MODE_TX ),
              .sample_rate = 44100, // corrected by info from bluetooth
              .bits_per_sample = (i2s_bits_per_sample_t) 16, /* the DAC module will only take the 8bits from MSB */
              .channel_format = I2S_CHANNEL_FMT_RIGHT_LEFT,
              .communication_format = I2S_COMM_FORMAT_I2S_MSB,
              .intr_alloc_flags = 0, // default interrupt priority
              .dma_buf_count = 8,
              .dma_buf_len = 64,
              .use_apll = false
              };
 static const i2s_pin_config_t pins = {.bck_io_num = 14,
                                     .ws_io_num = 15,
                                     .data_out_num = 27,
                                     .data_in_num = I2S_PIN_NO_CHANGE};
a2dp_sink.set_pin_config(pins);
a2dp_sink.set_i2s_config(i2s_config);
a2dp_sink.start("Supichaya");
}

void loop() {
}
```

**Comment by MichaelPNolan** [comment]
Ok so if I combine this with 
I2SStream out;
they do co-exist happily as long as I have something like this function to only use the copier to the i2s when i'm actively generating tones with other functions from:
#include "AudioTools.h"
Then Bluetooth doesn't stutter or have problems because it has the adequate processing power to service the audio stream. My code is currently using an effect stream to turn notes on and off so if it's not active then there is no sound from the sineWave generator that needs to be copied to the I2s stream.

```
    if(adsr.isActive()){
       copier.copy();
    } 
```

**Comment by pschatzmann** [comment, marked as answer]
I am not sure what you mean with 2 things can co-exist and it really depends on what you want to do.

Please note that A2DP is using up quite some RAM and some users where not able to implement TTS with their specific framework, though I managed to have it working with Tensorflow Light.

If you want to do some additional I2S output, it is better to switch off the automatic management by this library and take control yourself.

You can easily start/stop with the help of the __set_connected (bool active)__ method, if you want to disconnect or by calling stop() or play() if you want to temporarily halt the output via AVRCP.

**Comment by MichaelPNolan** [comment]
I'm just getting a feel for how things interact. Yes, i can see the Bluetooth player lib is using a lot of memory. Still, if it isn't active there is probably a lot of room left to do other things. I'm improvising but what I was thinking is to use this TDA7439 which needs a micro-controller to manage it. Then push the bluetooth into one of it's inputs - its a multi-device audio switcher for older type CD/Tape/Aux. 
So put this together because that was a project ... and maybe some kind of speech synthesis or http file player if that is possible.

Yes, probably not much point using bluetooth together with a synth but for that what I want is to make the synthesizer bluetooth midi enabled and that would be a different library. I'm not trying to necessarily take up more of your time - this is just the chat and show/tell so I hope you don't mind if this is just creative ideas banter.

**Comment by pschatzmann** [comment]
My [midi library](https://github.com/pschatzmann/arduino-midi) supports midi over BLE and Bluetooth Serial. Please note that if you use A2DP you can't use any other Bluetooth or WIFI functionality.

I have some synth examples with audio output to A2DP, but the lag is too big to be useful. 

Instead of the TDA7439 you can try the software solution: Here is [an example](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-audiokit/basic-a2dp-eq-audiokit/basic-a2dp-eq-audiokit.ino)

**Reply by MichaelPNolan** [reply]
It's possible I could use your eq and the TDA7439 since I have it anyway and it's been waiting for me to use. To eq the bluetooth input separately and show the parameters on the 1602 and put it in a nice box is just on my very long checklist of things to accomplish. 
What I mean is I have a lot of musical input/ouputs that would be good to deal with via that chip and so it's still worth looking at.

Thanks for pointing out the new possibilities using your library. 
Regards
Michael
